staging/src/k8s.io/client-go/tools/cache/synctrack/lazy.go

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package synctrack
+
+import (
+	"golang.org/x/sync/singleflight"
+	"sync/atomic"
+)
+
+// Lazy defers the computation of `Evaluate` to when it is necessary. It is
+// possible that Evaluate will be called in parallel from multiple goroutines.
+type Lazy[T any] struct {
+	Evaluate func() (T, error)
+
+	counter int64
+	cache   atomic.Value
+	group   singleflight.Group
+}
+
+type cacheEntry[T any] struct {
+	counter int64
+	result  T
+}
+
+// Notify should be called when something has changed necessitating a new call
+// to Evaluate.
+func (z *Lazy[T]) Notify() {
+	atomic.AddInt64(&z.counter, 1)
+	// This guarantees that a single caller who calls Notify and then Get
+	// in serial will not see an old cached value.
+	z.group.Forget("")
+}
+
+// Get should be called to get the current result of a call to Evaluate. If the
+// current cached value is stale (due to a call to Notify), then Evaluate will
+// be called synchronously. If subsequent calls to Get happen, they will all
+// get the same return value (due to use of singleflight).
+func (z *Lazy[T]) Get() (T, error) {
```

Comment:
I'm sure this code works but that seems excessively complicated. Why didn't you use a mutex rather than a combination of singleflight, atomic int64, and atomic value? Since eventually singleflight is still implemented with a mutex anyway?

Reply:
I've now pushed 2 commits; "if we use a lock instead" demonstrates the naive lock impl. The last commit shows the current solution but with singleflight rewritten out (fun exercise).

I also added a throughput test. The naive lock impl takes 7x as long as the impl in the final commit.

Reply:
oh, with the old singleflight impl, it took ~7.7s to run the tests 1000 times (sans throughput test). The current impl is a bit faster at ~7.1s.